src/gmail.js

```diff
@@ -2019,12 +2082,23 @@ var Gmail_ = function(localJQuery) {
     };
 
 
-    api.get.email_source_async = function(email_id, callback, error_callback) {
-        var url = api.helper.get.email_source_pre(email_id);
+    api.get.email_source_async = function(email_id, callback, error_callback, preferBinary) {
+        api.get.email_source_promise(email_id, preferBinary)
+            .then(callback)
+            .catch(error_callback);
+    };
+
+    api.get.email_source_promise = function(email_id, preferBinary) {
+        const url = api.helper.get.email_source_pre(email_id);
         if (url !== null) {
-            api.tools.make_request_async(url, "GET", callback, error_callback, true);
+            return api.tools.make_request_download_promise(url, preferBinary);
         } else {
-            callback("");
+            return new Promise((resolve, reject) => {
+                // I honestly don't think this makes sense.  We really
+                // should reject the promise, but that would break
+                // backward compatibility W.R.T. _async...
```

Comment:
If I were to be retro-critical to my own commit, and you agree we should aim for the more proper reject-semantics in an upcoming release anyway...

I think we could start aligning towards that goal right now by doing the right thing(tm) in the new `_promise` APIs (while leaving `_async` instact, for compatibility). How about it?